crda/remotematcher.go

```diff
@@ -0,0 +1,386 @@
+}
+
+// Configure implements driver.MatcherConfigurable.
+func (m *Matcher) Configure(ctx context.Context, f driver.MatcherConfigUnmarshaler, c *http.Client) error {
+	var cfg Config
+	if err := f(&cfg); err != nil {
+		return err
+	}
+
+	if cfg.Concurrency > 0 {
+		m.requestConcurrency = cfg.Concurrency
+	}
+
+	if cfg.URL != "" {
+		u, err := url.Parse(cfg.URL)
+		if err != nil {
+			return err
+		}
+		m.url = u
+	}
+	if cfg.Source != "" {
+		m.source = cfg.Source
+	}
+	if cfg.Key != "" {
+		m.key = cfg.Key
+	}
+	m.client = c
+
+	return nil
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "crda/Matcher.QueryRemoteMatcher"))
+	zlog.Debug(ctx).
+		Int("records", len(records)).
+		Msg("request")
+
+	// map Packge{name@version} to Packge to associate it with Vulnerability.
+	packageVersionToIndexRecord := make(map[string]*claircore.IndexRecord)
+	key := func(Name, Version string) string {
+		return fmt.Sprintf("%s@%s", Name, Version)
+	}
+	for _, ir := range records {
+		packageVersionToIndexRecord[key(ir.Package.Name, ir.Package.Version)] = ir
+	}
+
+	results := make(map[string][]*claircore.Vulnerability)
+	ctrlC := m.invokeComponentAnalysesInBatch(ctx, records)
+
+	for vrs := range ctrlC {
+		for _, vr := range vrs {
+			ir := packageVersionToIndexRecord[key(vr.Name, vr.Version)]
+
+			// A package can have 0 or more vulnerabilities for a version.
+			var vulns []*claircore.Vulnerability
+			for _, vuln := range vr.Vulnerabilities {
+				vulns = append(vulns, &claircore.Vulnerability{
+					ID:                 vuln.ID,
+					Updater:            "CodeReadyAnalytics",
+					Name:               vuln.ID,
+					Description:        vuln.Title,
+					Links:              vuln.URL,
+					Severity:           vuln.Severity,
+					NormalizedSeverity: normalizeSeverity(vuln.Severity),
+					FixedInVersion:     strings.Join(vuln.FixedIn, ", "),
+					Package:            ir.Package,
+					Repo:               ir.Repository,
+				})
+			}
+			results[ir.Package.ID] = append(results[ir.Package.ID], vulns...)
+		}
+	}
+	zlog.Debug(ctx).
+		Int("vulnerabilities", len(results)).
+		Msg("response")
+	return results, nil
+}
+
+func (m *Matcher) invokeComponentAnalysesInBatch(ctx context.Context, records []*claircore.IndexRecord) <-chan []*VulnReport {
+	ctrlC := make(chan []*VulnReport, len(records))
+	batchSize := m.batchSize
+	defer close(ctrlC)
+	var g errgroup.Group
+	for start := 0; start < len(records); start += batchSize {
+		start := start
+		end := start + batchSize
+		if end > len(records) {
+			end = len(records)
+		}
+		g.Go(func() error {
+			vulns, err := m.invokeComponentAnalyses(ctx, records[start:end])
+			if err != nil {
+				// log error
+				return nil
+			}
+			ctrlC <- vulns
+			return nil
+		})
+	}
+	g.Wait()
```

Comment:
Why don't we want to iterate through the channel here and return `[]*VulnReport`?

Reply:
@crozzy I did try doing that as per your reccomendation but the code erros out. Since ctrlC is still waiting for some api calls to return data its not possible to read fro ctrC and write into another variable.

Reply:
even after `g.Wait()` has been called? All goroutines should be done at that point, you should be able to close the ctrlC and iterate through the results. What was the error?

Reply:
`libvulnhttp     | panic: close of closed channel` This was the error.It has been resolved and code updated